enhancements/operator-sdk/downstream-operator-sdk.md

```diff
@@ -0,0 +1,391 @@
+  * As an Ansible operator developer, I would want to continue having a released downstream Ansible operator image I could base my operator on.
+  * As a Helm operator developer, I would want to continue having  a released downstream Helm operator image I could base my operator on.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Repos
+
+Mirror the upstream Operator SDK repos downstream in the openshift GitHub
+organization. We will have a one to one upstream to downstream repo.
+
+* operator-sdk - github.com/openshift/operator-sdk
+* operator-lib - github.com/openshift/operator-lib
+* ocp-release-operator-sdk - to be retired after the above repos are working
+
+The downstream repo will contain a special _overlay_ branch that contains all of
+the downstream-specific changes. The files to be contained in this overlay are
+as follows:
+
+* Dockerfiles to support CI
+* Dockerfiles to build downstream images
+* Vendor Go dependencies
+* Ansible collections to support Ansible operator
+
+We will pin the `downstream:master` to `upstream:master` and
+_upstream_release_branches_ to _downstream_release_branches_. This means we will
+manage the OpenShift release branches ourselves. So `release-x.y` will track a
+specific upstream branch that was mirrored. Only tagged commits will ever be
+brought into the release branches.
+
+#### Syncing
+
+Today, we sync `operator-framework/operator-sdk` into
+`openshift/ocp-release-operator-sdk`. We sync it manually with `git merge`. The
+`master` branch downstream currently tracks the latest OpenShift release. So
+when moving to a new major SDK release downstream, the `master` branch sync is
+frought with conflicts that can cause missing changes or just make for a very
+tedious merge.
+
+The idea behind the autosyncing strategy is that the downstream branches will
+match the upstream branches 1 for 1. So downstream `master` will track upstream
+`master`. Downstream `vx.y.z` will track upstream `vx.y.z`. So the syncs should
+occur with minimal conflicts. These downstream branches will NOT contain any
+special downstream fixes, they are strictly mirrors of the upstream.
+
+We will manage our own OpenShift release branches. These branches will be a
+combination of a tagged upstream branch plus the overlay branch.
+
+There is an initial prototype of the syncing code already:
+
+https://github.com/fabianvf/ocp-osdk-downstream-automation
+
+#### Overlay Branch Directory Structures
+
+Today, we have the following files in the `ocp-release-operator-sdk` repo. This
+usually doesn't impact the merge but it is a set of files that are NOT upstream.
+So we will put these in an overlay branch that can be used to lay down over the
+mirrored upstream branch to create a downstream release branch.
+
+The following is a list of directories and files that will exist in the overlay
+branch:
+
+* ci
+  * dockerfiles
+    * variety of Dockerfiles used to build images for running tests
+* tests
+  * e2e test scripts
```

Comment:
what means this tests dir? what tests will be added here?

I understand that the ci and test in https://github.com/openshift/ocp-release-operator-sdk/tree/master/ci/tests structure exist because it is how the SDK was in the past. The ocp-release-operator-sdk/ci/tests/ shows equivalent to operator-sdk/hack/tests/. Would not be easier to keep the same structure of sdk/upstream? Have we not an effort to move forward the tests for another directory?

also, downstream has https://github.com/openshift/ocp-release-operator-sdk/tree/master/ci/tests/scaffolding. Should we need to have the same for 1.0.0? Have we any reason to justify we have additional tests on downstream? Could we not use the sdk/upstream tests running against OCP only to ensure its features?

Reply:
Does this mean that for tests in downstream we will use OpenShift cluster? What would happen in the case where - we need a fix to run a feature in OCP but we do not face any issue locally(ex on minikube). For example, consider the issue: https://github.com/operator-framework/operator-sdk/issues/3701. The feature works well downstream when tested on Minikube, but  since the specific OCP version supports only `v1` CRDs we face errors. There could be scenarios where we may not require any changes downstream, but may need additional steps for running on ocp cluster. I am not sure if such a situation occurs often, but if that's the case do we need those downstream - since as a `Red Hat operator customer` I need not necessarily use OCP to run my operator.

Reply:
Hi @varshaprasad96, 

you are right. We have scenarios where issues were faced for OCP. However, is a pre-requirement to ensure the tool in the OCP cluster for downstream. We just not todo that in upstream as well to not compromise the CI. I mean, would take to long run the tests in OCP and kind in upstream for all PR's.

Reply:
You are correct, we should use the upstream tests but run them downstream. It is still unknown what needs to change to get the tests to run downstream.

Reply:
I added a story and removed the `tests` directory from the overlay.